code/modules/atmospherics/machinery/components/binary_devices/temperature_pump.dm

```diff
@@ -39,11 +39,28 @@
 	var/datum/gas_mixture/air_input = airs[1]
 	var/datum/gas_mixture/air_output = airs[2]
 
-	if((air_output.temperature + heat_transfer_rate) >= air_input.temperature || (air_input.temperature - heat_transfer_rate) <= TCRYO)
+	var/input_moles = air_input.total_moles()
+	var/output_moles = air_output.total_moles()
+
+	var/input_heat_capacity = air_input.heat_capacity()
+	var/output_heat_capacity = air_output.heat_capacity()
+
+	var/input_temperature = air_input.temperature
+	var/output_temperature = air_output.temperature
+
+	var/flux_rate = ((input_temperature - output_temperature) * heat_transfer_rate) * min(input_moles, output_moles)
+	//equalization check
+	flux_rate = min(flux_rate, ((input_temperature * input_heat_capacity * input_moles) - (output_temperature * output_heat_capacity * output_moles)) / 2)
+	if(flux_rate <= 0)
+		return
+	var/input_temp_change = -(input_heat_capacity / flux_rate)
+	var/output_temp_change = output_heat_capacity / flux_rate
+
+	if(input_temperature - input_temp_change < TCRYO)
 		return
 
-	air_input.temperature -= heat_transfer_rate
-	air_output.temperature += heat_transfer_rate
+	air_input.temperature += input_temp_change
+	air_output.temperature += output_temp_change
 	update_parents()
```

Comment:
I haven't checked this math yet, but the flow on this is ridiculously slow. did you test it?

Reply:
the math checks out, but i'm having some trouble with it, i'm doing some testing and check if fixes are needed